### PR TITLE
Feature/adhithya christen/firebase auth, sign up routes 

### LINKED ISSUE #27: Create Sign-up route
- [ ] Sign-up route should be created for the frontend to create users with. 
- [ ] Mentor/Mentee models should be implemented as discussed in DB Schema design #17 
- [ ] Add error handling (https://github.com/TritonSE/LAK-Goods-Transport-Application/tree/main/backend)

### PR DESCRIPTION
**Pull request by christen03**
### Tracking Info

Resolves #27 

Make sure your branch name conforms to: `<feature|staging|bugfix|...>/<username>/<3-4 word description separated by dashes>`. Otherwise, please rename your branch and create a new PR.

### Changes

We implemented POST routes for both mentee and mentor with error handling. We also connected our routes to Firebase, so for every user created, Firebase will also store their data along with a personal access token.

### Testing

We ran test requests to Postman and checked to ensure the data was actually being posted on MongoDB, which it was.

### Confirmation of Change 

Validation functions for routes can be found in the middleware folder **validation.ts**
Error handling changes can be found in the errors folder, **errors.ts, validationError.ts.**
Firebase connections can be found in the services folder, **auth.ts, firebase.ts**
Routes can be found in the routes folder, **routes.ts**
Updated mentee/mentor document models can be found in models, **mentee.ts, mentor.ts**

### CODE REVIEW

File: backend/src/config.ts
Comment (adhi0331) [inline backend/src/config.ts:42]: I had to add this thing to work around the linting error. Not an important issue right now but if there is a better workaround this, that would be nice because it is a bit messy.

File: backend/src/models/mentor.ts
Comment (adhi0331) [inline backend/src/models/mentor.ts:26]: After looking at Figma I realize that we need to update mentor and mentee schema to store additional values (images, major, interests, etc.). Not important now but it will be updated in future PR's.

**Review by adhi0331 — APPROVED** [review]
The route works for me properly. The error checking messages and mentor/mentee schema will need to be updated, but the route is good to be used for developing the sign-up flow on front-end.

**Review by HarshGurnani — APPROVED** [review]
All looks good to me, there seems to be lint errors in user.ts but if those console messages are supposed to be present then everything is good!

File: ALUM/ALUM.xcodeproj/project.pbxproj
Comment (AmanKAggarwal) [inline ALUM/ALUM.xcodeproj/project.pbxproj:71]: what is this?

File: backend/src/config.ts
Comment (AmanKAggarwal) [inline backend/src/config.ts:18]: Noting to make an issue for this task

File: backend/src/config.ts
Comment (AmanKAggarwal) [reply backend/src/config.ts:42]: Gotcha! Note to resolve this in the clean backend setup task we will make

File: backend/src/errors/validationError.ts
Comment (AmanKAggarwal) [inline backend/src/errors/validationError.ts:12]: An issue I faced with this approach in LAKTAA was that as we developed more, we had a lot of error messages so maintaining all of them like this became a pain. Maybe something to note for future

File: backend/src/middleware/validation.ts
Comment (AmanKAggarwal) [inline backend/src/middleware/validation.ts:4]: Add link to caketype - https://github.com/justinyaodu/caketype

File: backend/src/middleware/validation.ts
Comment (AmanKAggarwal) [inline backend/src/middleware/validation.ts:101]: Let's add the email check for mentor as well. See if you can reuse some logic between validateMentee and validateMentor

File: backend/src/routes/routes.ts
Comment (AmanKAggarwal) [inline backend/src/routes/routes.ts:1]: will we need this commented out code in future?

File: backend/src/routes/user.ts
Comment (AmanKAggarwal) [inline backend/src/routes/user.ts:20]: image

There are several fields still pending from the figma. This route will be called when the submit button is pressed and it will send all the data collected in the sign-up flow to here. Can you add those fields to the route and model?

File: backend/src/models/mentor.ts
Comment (AmanKAggarwal) [reply backend/src/models/mentor.ts:26]: It is kinda important so that the the frontend can be properly integrated with the Signup flow. Since this shouldn't take too long, can we add the fields?

File: backend/src/routes/user.ts
Comment (AmanKAggarwal) [inline backend/src/routes/user.ts:52]: Similar message for here. Check this figma for missing fields on this route.
image

File: backend/src/services/auth.ts
Comment (AmanKAggarwal) [inline backend/src/services/auth.ts:1]: Awesome!! Creating a service for auth and firebase is a great idea. These organizational things might seem not important right now but as our codebase increases in size, its usefulness will become clearer.

**Review by AmanKAggarwal — CHANGES_REQUESTED** [review]
Great work setting up all the auth stuff! Left some comments on finishing up the sign-up route but other than that looks good. Please ping on slack when you address the changes so that I can approve

File: backend/src/routes/routes.ts
Comment (christen03) [reply backend/src/routes/routes.ts:1]: so right, we don't need this anymore

File: backend/src/routes/user.ts
Comment (christen03) [reply backend/src/routes/user.ts:20]: add grade, topics of interest, career interests, and what do you hope question

File: backend/src/middleware/validation.ts
Comment (AmanKAggarwal) [reply backend/src/middleware/validation.ts:102]: ignore for now to unblock frontend! we can address this later

**Review by AmanKAggarwal — APPROVED** [review]
Great work! Let's merge this PR. There are some tasks pending on the mentor/mentee sign-up that we need to resolve as backlog. Can you make 1 task to resolve all this before merging?
- Add email regex check for mentor as well
- See if logic can be reused between mentor and mentee routes
- Add fields from figma so that Sign-Up Flow is done on backend

File: backend/src/models/mentor.ts
Comment (AmanKAggarwal) [reply backend/src/models/mentor.ts:26]: to be resolved - https://github.com/TritonSE/ALUM-Mobile-Application/issues/46

File: backend/src/config.ts
Comment (AmanKAggarwal) [reply backend/src/config.ts:18]: https://github.com/TritonSE/ALUM-Mobile-Application/issues/47

File: backend/src/config.ts
Comment (AmanKAggarwal) [reply backend/src/config.ts:42]: https://github.com/TritonSE/ALUM-Mobile-Application/issues/47

File: backend/src/routes/user.ts
Comment (AmanKAggarwal) [reply backend/src/routes/user.ts:20]: https://github.com/TritonSE/ALUM-Mobile-Application/issues/46

File: backend/src/routes/user.ts
Comment (AmanKAggarwal) [reply backend/src/routes/user.ts:52]: https://github.com/TritonSE/ALUM-Mobile-Application/issues/46

**Pull request merged**
